### PR TITLE
test/cql-pytest: tests CREATE/DROP INDEX during paged query

### PR DESCRIPTION
**Pull request by nyh**
This patch includes extensive testing for what happens to an ongoing
paged query when a secondary index is suddenly added or dropped.
Issue #18992 was opened suggesting that this would be broken, and indeed
the tests included here show that it is indeed broken.
    
The four tests included in this patch are heavily commented to explain
what they are testing and why, but here is a short summary of what is
being tested by each of them:
    
 1. A paged query filtering on v=17 continues correctly even if an
    index is created on v.
    
 2. A paged query filtering on v1 and v2 where v2 is indexed,
    continues correctly even if an index is created on v1 (remember
    that Scylla prefers to use the first index mentioned in the query).
    
 3. A paged query using an index on v continues correctly even if that
    index is deleted.
    
 4. However, if the query doesn't say "ALLOW FILTERING", it cannot
    be continued after the index is deleted.
    
All these tests pass on Cassandra, but all of them except the fourth
fail on Scylla, reproducing issue #18992. Somewhat to my suprise, the
failure of the query in all the failed tests is silent (i.e., trying to
fetch the next page just fetches nothing and says the iteration is done).
I was expecting more dramatic failures ("marshaling error" messages,
crashes, etc.) but didn't get them.
  
Refs #18992

    Signed-off-by: Nadav Har'El <nyh@scylladb.com>

### CODE REVIEW

**Comment by nyh** [comment]
Note to reviewers: The first patch in this PR is just https://github.com/scylladb/scylladb/pull/18993, which didn't reach master yet. Please ignore it in your review. I'll rebase and remove the first patch when #18993 goes into master.

@avikivity @michoecho these tests explore in depth the bug that you guys predicted - adding or removing an index in the middle of a paged query. It is indeed a real bug. I would argue a low priority bug (especially due to https://github.com/scylladb/scylladb/issues/7963), but it's a bug, and now we have tests to demonstrate it well.

**Review by avikivity — APPROVED** [review]

**Comment by nyh** [comment]
Rebased to get rid of the irrelevant first commit. This PR now has just one commit.

**Comment by denesb** [comment]
3 days later the CI is still running. Actually, it failed but didn't comment on the issue. Re-kicked.

File: test/cql-pytest/test_secondary_index.py [hunk 1]
```diff
@@ -1792,3 +1792,167 @@ def test_unbuilt_index_not_used(cql, test_keyspace, cassandra_bug):
...
+            r = cql.execute(stmt, paging_state=r.paging_state)
+            assert len(r.current_rows) <= page_size # sanity check
+            got.extend(r.current_rows)
+        assert expected == got
+
+# This test is the same as the previous one, but in this test the request
+# filters on both v1 and v2 is already using an index for column v2, and
+# now between the pages we add an index for v1. Reproduces #18992.
+@pytest.mark.xfail(reason="issue #18992")
+def test_paging_and_create_index2(cql, test_keyspace):
```
Comment (ptrsmrn) [inline test/cql-pytest/test_secondary_index.py:1847]: Maybe there's a better name for this function?

**Review by ptrsmrn — APPROVED** [review]

File: test/cql-pytest/test_secondary_index.py (hunk 1 above)
Comment (nyh) [reply test/cql-pytest/test_secondary_index.py:1847]: Do you have a good suggestion?
As I explained in the detailed comment, this test "is the same as the previous one, but...". I didn't want to have a 100-character function name explaining this "but" so I just used "2" :-)  I honestly think that a comment serves the readers better than a 100-character function name.

**Pull request closed without merging**